mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -147,34 +212,197 @@ fn get_addr_test_struct_field(variant: &syn::VariantData) -> Vec<quote::Tokens>
     res
 }
 
-/// This function generates a tuple which consists the definition of the struct and the 
-/// definition of the slot number.
-fn get_struct_def(name: &quote::Ident, variant: &syn::VariantData) -> (Vec<quote::Tokens>, Vec<quote::Tokens>) {
+lazy_static! {
+    static ref JSVAL: std::collections::HashMap<syn::Ident, syn::Ident> = {
+        let mut m = std::collections::HashMap::new();
+        m.insert(syn::Ident::from("f64"), syn::Ident::from("DoubleValue"));
+        m.insert(syn::Ident::from("u32"), syn::Ident::from("UInt32Value"));
+        m.insert(syn::Ident::from("i32"), syn::Ident::from("Int32Value"));
+        m.insert(syn::Ident::from("JSString"), syn::Ident::from("StringValue"));
+        m.insert(syn::Ident::from("bool"), syn::Ident::from("BooleanValue"));
+        m.insert(syn::Ident::from("JSObject"), syn::Ident::from("ObjectValue"));
+        m
+    };
+}
+
+/// This function generates a tuple which consists the definition of the struct, the
+/// definition of the slot number, getters and setters for the object, js accesible
+/// getter and setters and JSPropertySpec array elements.
+fn get_struct_def(name: &quote::Ident,
+                  variant: &syn::VariantData)
+                  -> (Vec<quote::Tokens>, Vec<quote::Tokens>, Vec<quote::Tokens>,
+                      Vec<quote::Tokens>, Vec<quote::Tokens>, Vec<quote::Tokens>,
+                      Vec<quote::Tokens>) {
```

Comment:
A couple nitpicks for readability/maintainability:

* Let's define a struct with field names instead of having to know the ordering of the tuples seven fields:

```rust
struct MagicStructCode {
    /// A definition of the magic struct itself.
    struct_def: Vec<quote::Tokens>,

    /// The definition of the number of slots the magic struct requires.
    slots: Vec<quote::Tokens>,

    /// Rust getters for the data stored in slots.
    getters: Vec<quote::Tokens>,

    /// Rust setters for the data stored in slots.
    setters: Vec<quote::Tokens>,

    /// `JSNative` getter functions exposed to JavaScript.
    js_getters: Vec<quote::Tokens>,

    /// `JSNative` setter functions exposed to JavaScript.
    js_setters: Vec<quote::Tokens>,

    /// The `JSPropertySpec` definition that exposes the `JSNative`s to JavaScript.
    js_prop_spec: Vec<quote::Tokens>,
}
```

* Can we pull generating each getter, setter, etc out into its own function that we call from inside the loop over each field? This function is a behemoth right now, which makes it hard to follow what is going on.